src/org/traccar/protocol/CarcellProtocolDecoder.java

```diff
@@ -0,0 +1,147 @@
+import org.traccar.helper.PatternBuilder;
+import org.traccar.helper.UnitsConverter;
+import org.traccar.model.Position;
+
+public class CarcellProtocolDecoder extends BaseProtocolDecoder {
+
+    public CarcellProtocolDecoder(CarcellProtocol protocol) {
+        super(protocol);
+    }
+
+    private static final Pattern PATTERN = new PatternBuilder()
+            .expression("([$%])")                // memory flag
+            .number("(d+),")                     // imei
+            .text("CEL,").optional()
+            .groupBegin()
+            .number("([NS])(dd)(dd).(dddd),")    // latitude
+            .or()
+            .number("([NS])(d+.d+),")            // latitude
+            .groupEnd()
+            .groupBegin()
+            .number("([EW])(ddd)(dd).(dddd),")   // longitude
+            .or()
+            .number("([EW])(d+.d+),")            // longitude
+            .groupEnd()
+            .number("(d+),")                     // speed
+            .number("(d+),")                     // course
+            .number("([-+]ddd)([-+]ddd)([-+]ddd),")       // x,y,z
+            .number("(d+),")                     // battery
+            .number("(d+),")                     // csq
+            .number("(d),")                      // jamming
+            .number("(d+),")                     // hdop
+            .expression("([CG])")                // clock type
+            .number("(dd)(dd)(dd),")             // date
+            .number("(dd)(dd)(dd),")             // time
+            .number("(d),")                      // block
+            .number("(d),")                      // ignition
+            .number("(d),")                      // cloned
+            .expression("([AF])")                // panic
+            .number("(d),")                      // painel
+            .number("(d+),")                     // battery voltage
+            .any()                               // full format
+            .compile();
+
+    @Override
+    protected Object decode(
+            Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+
+        String sentence = (String) msg;
+
+        Parser parser = new Parser(PATTERN, sentence);
+
+        if (!parser.matches()) {
+            return null;
+        }
+
+        Position position = new Position();
+        position.setProtocol(getProtocolName());
+        position.set(Position.KEY_ARCHIVE, parser.next().equals("%"));
+        position.setValid(true);
+
+        if (!identify(parser.next(), channel, remoteAddress)) {
+            return null;
+        }
+
+        position.setDeviceId(getDeviceId());
+
+        if (sentence.indexOf("CEL,") < 0) {
+            position.setLatitude(parser.nextCoordinate(CoordinateFormat.HEM_DEG_MIN_MIN));
+            parser.next();
+            parser.next();
+            position.setLongitude(parser.nextCoordinate(CoordinateFormat.HEM_DEG_MIN_MIN));
+            parser.next();
+            parser.next();
```

Comment:
I would combine both coordinates in a single group. Then instead of a wall of `next()` calls and a weird check for `CEL` at the beginning, you can use `hasNext(N)`.